File: UnitTest/Zeze/Trans/TestConflict.cs
﻿using System;
using System.Collections.Generic;
using System.Text;
using System.Threading.Tasks;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Zeze.Transaction;
using Zeze.Util;

namespace UnitTest.Zeze.Trans
{
    [TestClass]
    public class TestConflict
    {
        int sum;

        [TestInitialize]
        public void TestInit()
        {
            demo.App.Instance.Start();
        }

        [TestCleanup]
        public void TestCleanup()
        {
            demo.App.Instance.Stop();
        }

        [TestMethod]
        public void TestConflictAdd()
        {
            Assert.IsTrue(ResultCode.Success == demo.App.Instance.Zeze.NewProcedure(ProcRemove, "ProcRemove").CallSynchronously());
            Task[] tasks = new Task[2000];
            for (int i = 0; i < tasks.Length; ++i)
            {
               tasks[i] = demo.App.Instance.Zeze.NewProcedure(ProcAdd, "ProcAdd").CallAsync();
            }
            Task.WaitAll(tasks);
            sum = tasks.Length;
            Assert.IsTrue(ResultCode.Success == demo.App.Instance.Zeze.NewProcedure(ProcVerify, "ProcVerify").CallSynchronously());
            Assert.IsTrue(ResultCode.Success == demo.App.Instance.Zeze.NewProcedure(ProcRemove, "ProcRemove").CallSynchronously());
        }

        async Task<long> ProcRemove()
        {
            await demo.App.Instance.demo_Module1.Table1.RemoveAsync(123123);
            return ResultCode.Success;
        }

        async Task<long> ProcAdd()
        {
            demo.Module1.Value v = await demo.App.Instance.demo_Module1.Table1.GetOrAddAsync(123123);
            v.Int_1 += 1;
            return ResultCode.Success;
        }

        async Task<long> ProcVerify()
        {
            demo.Module1.Value v = await demo.App.Instance.demo_Module1.Table1.GetOrAddAsync(123123);
            Assert.IsTrue(v.Int_1 == sum);
            return ResultCode.Success;
        }
    }
}

File: ZezeJava/ZezeJavaTest/src/UnitTest/Zeze/Trans/TestConflict.java
package UnitTest.Zeze.Trans;

import java.util.ArrayList;
import java.util.concurrent.Future;
import demo.Module1.BValue;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import Zeze.Transaction.Procedure;

public class TestConflict {
	private int sum;

	@BeforeEach
	public final void testInit() throws Exception {
		demo.App.getInstance().Start();
	}

	@AfterEach
	public final void testCleanup() throws Exception {
		//demo.App.getInstance().Stop();
	}

	@Test
	public final void testConflictAdd() throws Exception {
		Assertions.assertEquals(Procedure.Success, demo.App.getInstance().Zeze.newProcedure(TestConflict::ProcRemove, "ProcRemove").call());
		var tasks = new ArrayList<Future<?>>();
		for (int i = 0; i < 2000; ++i) {
			tasks.add(Zeze.Util.TaskSpec.ofProcedure(
					demo.App.getInstance().Zeze.newProcedure(TestConflict::ProcAdd, "ProcAdd")).submitNow());
			if ((i + 1) % 200 == 0) {
				for (Future<?> task : tasks)
					task.get();
				sum += tasks.size();
				tasks.clear();
			}
		}
		for (Future<?> task : tasks) {
			task.get();
		}
		sum += tasks.size();
		Assertions.assertEquals(Procedure.Success, demo.App.getInstance().Zeze.newProcedure(this::ProcVerify, "ProcVerify").call());
		Assertions.assertEquals(Procedure.Success, demo.App.getInstance().Zeze.newProcedure(TestConflict::ProcRemove, "ProcRemove").call());
	}

	private static long ProcRemove() {
		demo.App.getInstance().demo_Module1.getTable1().remove(123123L);
		return Procedure.Success;
	}

	private static long ProcAdd() {
		BValue v = demo.App.getInstance().demo_Module1.getTable1().getOrAdd(123123L);
		v.setInt_1(v.getInt_1() + 1);
		return Procedure.Success;
	}

	private long ProcVerify() {
		BValue v = demo.App.getInstance().demo_Module1.getTable1().getOrAdd(123123L);
		Assertions.assertEquals(v.getInt_1(), sum);
		return Procedure.Success;
	}
}
